### PR TITLE
ui: highlight sidebar links for nested pages

### LINKED ISSUE #24069: ui: highlight database sidebar link when on table details page
The sidebar link for the databases page is not highlighted when on the table details page, because the URL for the index page is `/databases` and the URL for the table details is `/database/:dbName/table/:tableName`.  We should do some teaking to the default logic to make this highlight in the way you'd expect.

### PR DESCRIPTION
**Pull request by couchand**
The route reorganization broke one of react-router's assumptions: that
logically nested pages are always on strictly-nested routes.  This
fixes the sidebar link highlighting for nested pages whose route
doesn't start with the exact route the link is going to.

Fixes: #24069
Release note: None

screen shot 2018-04-11 at 3 46 24 pm

### CODE REVIEW

**Comment by couchand** [comment]
@vilterp one question I had here was if we should just use this facility for all highlighting, rather than splitting the logic between react-router for the default behavior and this custom logic for the override.  It might be a bit simpler/more obvious to only use this and never rely on react-router at all.  What do you think?

File: pkg/ui/src/views/app/components/layoutSidebar/index.tsx [hunk 1]
```diff
@@ -49,10 +71,10 @@ export default class extends React.Component<{}, {}> {
   render() {
     return <nav className="navigation-bar">
       <ul className="navigation-bar__list">
-        <IconLink to="/overview" icon={homeIcon} title="Overview" />
+        <IconLink to="/overview" icon={homeIcon} title="Overview" indexFor={["/nodes", "/node"]} />
```
Comment (vilterp) [inline pkg/ui/src/views/app/components/layoutSidebar/index.tsx]: A comment on `IconLinkProps` explaining the semantics of `indexFor` would be nice — I assume that it does a prefix match, so e.g. `/nodes/5` will cause the `/overview` link to be highlighted? Unclear that that would work from [this SO answer about `router.isActive`](https://stackoverflow.com/questions/43750056/checking-router-isactive-on-nested-routes-in-react?utm_medium=organic&utm_source=google_rich_qa&utm_campaign=google_rich_qa)

**Review by vilterp — APPROVED** [review]
Interesting question about using react-router's `activeClassName` thing, which relies on the nested-routes assumption we're breaking, vs. this new thing you came up with.

I guess my opinion is that I wish we weren't breaking that assumption — that there actually was a `nodes/` index and a `nodes/:id`. But that ship has sailed — the overview is the node list now (as well as showing other non-node things like ranges), and we've removed the standalone node list. Maybe some way of solving this will occur to us as part of our IA discussion.

Until we think of an elegant solution, I think what you have now is fine. I'd just request that you add a little comment explaining why `indexFor` is necessary (something like what you have in the PR description) and how its semantics (left a comment about the specific part that confused me).

File: pkg/ui/src/views/app/components/layoutSidebar/index.tsx (hunk 1 above)
Comment (couchand) [reply pkg/ui/src/views/app/components/layoutSidebar/index.tsx]: yes, I spent a little while trying to figure out an obvious name for the prop... just changed it to `activeFor`, do you think that's clear enough, particularly after unifying all active class logic?

File: pkg/ui/src/views/app/components/layoutSidebar/index.tsx [hunk 2]
```diff
@@ -1,4 +1,7 @@
+import classnames from "classnames";
```
[the remark this answers is not in the document]
Comment (couchand) [reply pkg/ui/src/views/app/components/layoutSidebar/index.tsx]: done

**Review by vilterp — APPROVED** [review]
LGTM.

Don't have a strong opinion here, but why `/node/<n>` instead of `/nodes/<n>`?

File: pkg/ui/src/views/app/components/layoutSidebar/index.tsx [hunk 3]
```diff
@@ -23,14 +26,25 @@ interface IconLinkProps {
 class IconLink extends React.Component<IconLinkProps, {}> {
   static defaultProps = {
     className: "normal",
-    onlyActiveOnIndex: false,
+    activeFor: [] as string[],
+  };
```
[the remark this answers is not in the document]
Comment (couchand) [reply pkg/ui/src/views/app/components/layoutSidebar/index.tsx]: good point

[the remark this answers is not in the document]
Comment (couchand) [reply pkg/ui/src/views/app/components/layoutSidebar/index.tsx]: Partial<IconLinkProps> to optionalize everything but otherwise great!

**Comment by couchand** [comment]
> why `/node/<n>` instead of `/nodes/<n>`?

Mainly to reserve the collection namespace for interesting sub-collections.  For instance, `/nodes/decommissioned` or some such.

**Comment by vilterp** [comment]
Interesting, hadn't thought of the reserving the namespace consideration.

**Pull request merged**